overlord/snapstate/snapstate_update_test.go

```diff
@@ -9285,25 +9285,23 @@ func (s *snapmgrTestSuite) TestDownloadTaskWaitsForPreDownload(c *C) {
 			c.Assert(s.o.TaskRunner().Ensure(), IsNil)
 
 			for i := 0; i < 5; i++ {
-				select {
-				case <-time.After(time.Second):
-					s.state.Lock()
-					atTime := dlTask.AtTime()
-					s.state.Unlock()
-					if atTime.IsZero() {
-						continue
-					}
+				<-time.After(time.Second)
+				s.state.Lock()
+				atTime := dlTask.AtTime()
+				s.state.Unlock()
+				if atTime.IsZero() {
+					continue
+				}
 
-					s.state.Lock()
-					defer s.state.Unlock()
+				s.state.Lock()
+				defer s.state.Unlock()
 
-					// the download task registers itself w/ the pre-download and retries
-					c.Assert(atTime.Equal(now.Add(2*time.Minute)), Equals, true)
-					var taskIDs []string
-					c.Assert(preDlTask.Get("waiting-tasks", &taskIDs), IsNil)
-					c.Assert(taskIDs, DeepEquals, []string{dlTask.ID()})
-					return
-				}
+				// the download task registers itself w/ the pre-download and retries
+				c.Assert(atTime.Equal(now.Add(2*time.Minute)), Equals, true)
+				var taskIDs []string
+				c.Assert(preDlTask.Get("waiting-tasks", &taskIDs), IsNil)
+				c.Assert(taskIDs, DeepEquals, []string{dlTask.ID()})
```

Comment:
If the assertion with the `Get` fails, then the next one stops making sense (and its output just confuses things) so IMO an `Assert` is appropriate. Regarding the `defer` in the loop, there's a `return` immediately after these checks so I think that clarifies that the `defer` isn't running more than once

> I only suggest this because I did once encounter an Assert inside this loop while doing the single reboot PR, and it made it fail in a weird way that made the error output almost undebuggable

How so? Is it the `panic: unlocked of an unlocked mutex` problem?

Reply:
Yes exactly, sorry for the slow response, it's the unlocked of an unlocked mutex

Reply:
No worries. That's a different issue and IMO it wouldn't be fixed by switching to a non-deferred `Unlock` in this code (it's functionally equal since the `Unlock` will happen anyway, even if the assert fails). The `unlock of an unlocked mutex` panic happens because we defer an `Unlock()` at the top of the test which then explodes if we stop running at some point where the lock isn't held. Sadly, this is a common issue in our tests but the stack trace should still have the necessary info to debug it (although a bit muddied up, as you pointed out).